### PR TITLE
Support userIdentifier to local userName mapping via uid

### PR DESCRIPTION
**Pull request by jeffvroom**
### Intent

In the pro version, it seems increasingly common that the assertion the auth-system provides is either an email address, upper-case user-name, or other alias for the system username. Although they can login using these aliases, RSW passes the uid to the session launch that then gets the canonical username. This creates problems for the file-name of the unix socket  and is a blocker for two customers.

Addresses: https://github.com/rstudio/rstudio-pro/issues/2736

### Approach

This PR takes the existing userIdentifierToLocalUsername function in ServerSAMLAuth.cpp and moves it into ServerAuthCommon.cpp so it can be used by all auth-handlers. Further it changes that method so it uses uid to fetch the username (since getpwnam can return the username it was passed, not the canonical one for that uid) 

A companion PR for pro will call this method for openId, SAML, and proxy auth handlers.

In this PR, this is always-on but I could add a new server option if anyone thinks that's valuable. 

### Automated Tests

The existing automated tests will verify that login works correctly with this change. If needed, I could add a user-name alias to one of our existing test configurations, and verify that we can start a session using this alias. 

### QA Notes

It's possible to do an adduser aliasName and then edit /etc/passwd to change the uid of aliasName to match an existing user. On ubuntu, you can create an email-address as a username by adding the --force-bad-names option to adduser. 

When you login using one of these aliases, the UI will show you logged in using the canonical name. 

### Checklist

- [TBD ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x ] This PR passes all local unit tests

### CODE REVIEW

**Review by ricardofandrade — APPROVED** [review]
Nice! So it means Linux itself can provide username aliasing (to say, email).
Makes sense because LDAP auth w/ PAM works this way for some customers.
I haven't thought about that before, but it makes sense!

File: src/cpp/server/ServerPAMAuth.cpp [hunk 1]
```diff
@@ -162,15 +130,15 @@ void doSignIn(const http::Request& request,
                                                             &plainText);
       if (error)
       {
-         LOG_ERROR(error);
+         LOG_ERROR_MESSAGE("Failed sign-in - unable to decrypt password - error: " + error.asString());
```
Comment (jmcphers) [inline src/cpp/server/ServerPAMAuth.cpp:133]: Here and elsewhere - I think we'll lose some of the error metadata if we don't log the error directly so we should add these as properties.
```suggestion
         error.addDescription("Failed sign-in - unable to decrypt password");
         LOG_ERROR(error);
```

**Review by jmcphers — APPROVED** [review]

File: src/cpp/server/auth/ServerAuthCommon.cpp [hunk 2]
```diff
@@ -424,6 +439,60 @@ void prepareHandler(handler::Handler& handler,
...
+      {
+         // log the error and return the original user identifier as a fallback
+         error.addProperty("description", "Error converting userIdentifier to username");
+         LOG_ERROR(error);
+      }
+      else
+      {
+         // This gets the passwd entry for the user-id - this is what the session will do so
+         // when a uid is aliased, we need to go back to the uid for the real username
+         error = core::system::User::getUserFromIdentifier(user.getUserId(), user);
```
Comment (MariaSemple) [inline src/cpp/server/auth/ServerAuthCommon.cpp:471]: Once you have the user object, haven't you already read the passwd entry? Unless I'm mistaken both `User::getUserFromIdentifier` calls should return a fully populated user object

Comment (jeffvroom) [reply src/cpp/server/auth/ServerAuthCommon.cpp:471]: In my testing, if you have an alias where there are two pwd entries for the same uid, getUserFromIdentifier returns the same username in the pw entry - i.e. the alias. So the second call makes sure that we get the same pw entry that the session will get.

**Pull request merged**